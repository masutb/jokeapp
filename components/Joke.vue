<template>
  <div class="item" :style="{ fontSize: fontSize + 'px' }">
    <span class="part" v-for="(part, i) in textParts" :key="i">{{ part + (i !== textParts.length - 1 ? '&nbsp;' : '')}}</span>
  </div>
</template>

<script>
import anime from 'animejs'

export default {
  props: {
    data: {
      type: Object,
      required: true
    },
    active: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    jokeText() {
      return this.data.type === 'twopart' ? this.data.setup + ' ' + this.data.delivery : this.data.joke
    },
    textParts() {
      if (!this.jokeText.length) {
        return ['Try again.']
      }

      return this.jokeText.split(' ')
    },
    fontSize() {
      if (!this.jokeText.length) {
        return 80
      }

      let size = 80 + ((80 - this.jokeText.length) / 7.5)
      return Math.max(size, 32)
    }
  },
  watch: {
    active(val) {
      this.toggle(val)
    }
  },
  mounted() {
    this.active && this.toggle()
  },
  methods: {
    toggle(active = true) {
      if (active) {
        setTimeout(() => {
          anime({
            targets: this.$el.children,
            translateY: [-100, 0],
            opacity: 1,
            duration: 600,
            delay: (el, i) => 15 * i
          })
        }, 300)
      } else {
        anime({
          targets: this.$el.children,
          translateY: [0, -100],
          opacity: 0,
          duration: 600,
          delay: (el, i) => 15 * i
        })
      }
    }
  }
}
</script>

<style lang="less" scoped>
.item {
  position: relative;
  display: inline-block;
  font-size: 60px;
  font-family: @font-fancy;
  color: #fff;
  margin: 0 auto;
  text-align: center;
  word-break: break-word;

  ::v-deep > span {
    transform: translateY(-100px);
    display: inline-block;
    opacity: 0;
  }
}
</style>
