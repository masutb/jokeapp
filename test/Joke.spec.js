import { mount } from '@vue/test-utils'
import Joke from '@/components/Joke.vue'

const data = {
  "error": false,
  "category": "Dark",
  "type": "single",
  "joke": "Foo.",
  "flags": {
    "nsfw": false,
    "religious": false,
    "political": false,
    "racist": false,
    "sexist": false
  },
  "id": 167,
  "lang": "en"
}

describe('Joke', () => {
  let wrapper = mount(Joke, {
    propsData: {
      data,
      active: true
    }
  })

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })
  
  test("renders properly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  })

  test("toggle is called when active prop changes", () => {
    const wrapper = mount(Joke, {
      propsData: {
        data,
        active: true
      }
    })

    const spyUpdate = jest.spyOn(wrapper.vm, 'toggle')

    wrapper.setProps({ active: false })

    wrapper.vm.$nextTick(() => {
      expect(spyUpdate).toHaveBeenCalled()
    })

    spyUpdate.mockReset()
  })

  test("toggle with true param shows joke words", () => {
    wrapper.vm.toggle(true)

    wrapper.vm.$nextTick(() => {
      expect(wrapper.findAll('.part').at(0).isVisible()).toBe(true)
    })
  })

  test("toggle with false param hides joke words", () => {
    wrapper.vm.toggle(false)

    wrapper.vm.$nextTick(() => {
      expect(wrapper.findAll('.part').at(0).isVisible()).toBe(false)
    })
  })

  test("renders \"Try again.\" text if joke text is empty", () => {
    data.joke = ''
    wrapper = mount(Joke, {
      propsData: {
        data,
        active: true
      }
    })
    expect(wrapper.text()).toContain('Try again.')
  })
})
