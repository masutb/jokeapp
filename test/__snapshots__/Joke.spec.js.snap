// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke renders properly 1`] = `"<div class=\\"item\\" style=\\"font-size: 90.13333333333333px;\\"><span class=\\"part\\">Foo.</span></div>"`;
